labs/optimizely-data-cli/oevents

```diff
@@ -0,0 +1,651 @@
+
+  $ $CLI_NAME paths
+  
+  s3://optimizely-events-data/v1/account_id=12345/
+
+  $ $CLI_NAME paths --type decisions
+
+  s3://optimizely-events-data/v1/account_id=12345/type=decisions/
+
+  $ $CLI_NAME paths --type decisions --date 2020-07-01
+
+  s3://optimizely-events-data/v1/account_id=12345/type=decisions/date=2020-07-01/
+
+  $ $CLI_NAME paths --type decisions --start 2020-07-01 --end 2020-07-05
+
+  s3://optimizely-events-data/v1/account_id=12345/type=decisions/date=2020-07-01/
+  s3://optimizely-events-data/v1/account_id=12345/type=decisions/date=2020-07-02/
+  s3://optimizely-events-data/v1/account_id=12345/type=decisions/date=2020-07-03/
+  s3://optimizely-events-data/v1/account_id=12345/type=decisions/date=2020-07-04/
+  s3://optimizely-events-data/v1/account_id=12345/type=decisions/date=2020-07-05/
+
+  $ $CLI_NAME paths --type decisions --start 2020-07-01 --end 2020-07-05 --experiment 56789
+
+  s3://optimizely-events-data/v1/account_id=12345/type=decisions/date=2020-07-01/experiment=56789/
+  s3://optimizely-events-data/v1/account_id=12345/type=decisions/date=2020-07-02/experiment=56789/
+  s3://optimizely-events-data/v1/account_id=12345/type=decisions/date=2020-07-03/experiment=56789/
+  s3://optimizely-events-data/v1/account_id=12345/type=decisions/date=2020-07-04/experiment=56789/
+  s3://optimizely-events-data/v1/account_id=12345/type=decisions/date=2020-07-05/experiment=56789/
+"
+
+################################################################################
+# Logging and Error output                                                     #
+################################################################################
+
+# VerboseEcho
+# Echo the supplied text if the --verbose flag is on
+function VerboseEcho()
+{
+  if [ "$verbose" = true ]; then 
+    echo "$@"
+  fi
+}
+
+# ErrorEcho
+# Echo the supplied text to stderr 
+function ErrorEcho()
+{
+  echo "$@" >&2
+}
+
+################################################################################
+# Prerequisites                                                                #
+################################################################################
+
+function Require()
+{
+  local requirement="$1"
+
+  if ! command -v "$requirement" &> /dev/null; then
+    ErrorEcho "Error: $CLI_NAME requires $requirement to run.  Please install it and try again."
+    exit 1
+  fi
+}
+
+
+################################################################################
+# Date and Time                                                                #
+################################################################################
+
+# TimestampMsToDateString
+# Convert a unix timestamp (in milliseconds) to a date string and return the result
+function TimestampMsToDateString()
+{
+  local timestamp_ms="$1"
+  local timestamp_s=$(( timestamp_ms / 1000 ))
```

Comment:
The reason for this is because assignment itself is a command with a dedicated exit status that will usually succeed even the operation fails, causing a bad signal if one is expecting to check status code.

Reply:
oh, wow. Confirmed via:

```sh
myfunc() {
    local var1=$(false); echo "$?" # 0

    local var2=
    var3=$(false); echo "$?" # 1 

    declare var3=$(false); echo "$?" # 0

    declare var4=
    var4=$(false); echo "$?" # 1

    var5=$(false); echo "$?" # 1
}
myfunc
```

So it seems that it's the explicit declaration via `local` or `declare`, not the assignment, which is executed after the command substitution (thus mucking up the exit code)

Reply:
Fixed.